### ISSUE TITLE
Errors and broken pdf when run with unprivileged user

### ISSUE THREAD

**Issue by ebogaard** [tesseract-ocr/tesseract#364]
When running tesseract as root like this:

`tesseract in.tiff out  -l nld+eng pdf`

The output is a proper and functioning pdf.
When I run the same command as an unprivileged user, like this:

`sudo -u userx tesseract in.tiff out  -l nld+eng pdf`

The output is as follows:

```
Tesseract Open Source OCR Engine v3.05.00dev with Leptonica
Page 1
Error in fopenWriteStream: stream not opened
Error in pixWrite: stream not opened
Error in fopenReadStream: file not found
Error in extractG4DataFromFile: stream not opened to file
Error in l_generateG4Data: datacomp not extracted
Error in pixGenerateCIData: g4 data not made
Error in l_generateCIDataForPdf: file in.tiff format is 4; unreadable
Error during processing.
```

The unprivileged user has access to both the input and the output file.
Tesseract is compiled with the following options:

```
sudo -u userx tesseract -v
tesseract 3.05.00dev
 leptonica-1.73
  libjpeg 6b (libjpeg-turbo 1.2.90) : libpng 1.5.13 : libtiff 4.0.3 : zlib 1.2.7
```

The mentioned libs are available on the system.
The user has a normal account with shell: userxx:1001:1001::/home/userx:/bin/bash

When I let tesseract output to txt (so I don't use the 'pdf'-option), there is no error and a txt-file with recognized text is output. This works both for privileged and unprivileged users.

**Comment by amitdo** [comment]
All the errors except the last one are coming from [leptonica](https://github.com/DanBloomberg/leptonica).

Check with a png image, just to make sure the source of the problem is not reading the tiff image.

**Comment by ebogaard** [comment]
Yeah, I found that out later as well. But I'm not sure where Tesseract stops and Leptonica begins ;-)

Previously I tried tiff and pbm as sources. Both gave the same errors; the only thing that changed then, was the one-before-last error: tiff gives a 'format is 4'- error while pbm gives a 'format is 11'-error.

Just tried it again: when I create a new source and make the type png it actually does work!
After converting this png to tiff, that does work as well, but converting to pbm (which is used by pdfsandwich, which I want to use) still gives the errors. I can open and view the bpm file without problems, though.

What do you make of this? Is this a problem with Tesseract, Leptonica, with the source file(s) or libraries?

**Comment by zdenop** [comment]
File input/output is handles by leptonica.
Anyway your experience is strange. I build tesseract (and leptonica) as non privileged user and install it with `sudo` and I never have this problem.
Can you please remove/uninstall leptonica and tesseract and than clone last version of leptonica and tesseract and build everything under userx , install with sude and try again your test?
If you have still problem can you provide your in.tiff file?

**Comment by ebogaard** [comment]
I'm afraid compiling, installing and then executing as userx doesn't improve the situation.
png and tiff are working, pbm still gives the same error:

```
Tesseract Open Source OCR Engine v3.05.00dev with Leptonica
Error in fopenWriteStream: stream not opened
Error in pixWrite: stream not opened
Error in fopenReadStream: file not found
Error in extractG4DataFromFile: stream not opened to file
Error in l_generateG4Data: datacomp not extracted
Error in pixGenerateCIData: g4 data not made
Error in l_generateCIDataForPdf: file test-source.pbm format is 11; unreadable
Error during processing.
```

I tried to make sure the pbm was converted right, so I used convert with and without the options '-compress none -depth 2'. Both files give the same problem.

**Comment by amitdo** [comment]
Try converting the image with other programs like gimp or pngtopnm (netpbm).

**Comment by ebogaard** [comment]
Converted the png with pngtopam, but this time got a pgm (P5, 8 bits) instead of a pbm (P4, 2 bits).
And that actually does work!
When I use convert instead of pngtopam, to create a pgm, this also works.

So it seems the problem wasn't really with using an unprivileged user, it's a problem with pbm as a source filetype.

Should we probably close this issue and open another about that?

**Comment by amitdo** [comment]
>Should we probably close this issue and open another about that?

As said before, Tesseract uses Leptonica for image IO, so if there is a problem with a certain image format there is noting Tesseract devs can do about it. So don't open another issue.

**Comment by amitdo** [comment]
Try [pamditherbw](http://netpbm.sourceforge.net/doc/pamditherbw.html) on the pgm file.

**Comment by jbreiden** [comment]
This bug report is confusing. Please generate shareable test images that give trouble, and post to this bug. Also, please report your platform details. Is this Linux or something else?

**Comment by ebogaard** [comment]
Okay, fair enough. Let's dig deeper.
This is on Centos 7. I try to use pdfsandwich to OCR existing PDFs, which uses a combination of ImageMagick + unpaper + tesseract + Ghostcript.

As Centos doesn't provide all packages or ones that are recent enough, I compiled unpaper 6.1, leptonica 1.73 and tesseract 3.05dev from source.

What does work:
- png to pdf
- tiff to pdf
- pbm to pdf => only when root, not as a regular, unprivileged user
- pgm to pdf (works from the command line, but unpaper has a problem with this type, but that's another problem)

See attached sourcefiles:

[sources.zip](https://github.com/tesseract-ocr/tesseract/files/369888/sources.zip)

**Comment by amitdo** [comment]
Since 2 out of 2 of my advices help you, maybe you should try my last advice... :)

**Comment by jbreiden** [comment]
Thanks. By the way, the "root-only" thing suggests that there may be some sort of tmpfile involved that is running into a permission problem. I'll be excited if I can reproduce.

**Comment by jbreiden** [comment]
Leptonica is trying to write its temporary files in /tmp/lept and can't because root owns it. This is going to require a Leptonica change.

```
sudo sudo -u nobody strace tesseract test-source2.pbm /tmp/q pdf
mkdir("/tmp", 0777)                     = -1 EEXIST (File exists)
mkdir("/tmp/lept", 0777)                = -1 EEXIST (File exists)
open("/tmp/lept/868647_30487_temp.tif", O_RDWR|O_CREAT|O_TRUNC, 0666) = -1 EACCES (Permission denied)
write(2, "Error in fopenWriteStream: strea"..., 45Error in fopenWriteStream: stream not opened
```

**Comment by ebogaard** [comment]
I can confirm the permissions on /tmp/lept are indeed the problem.
When I do 'chmod 777 /tmp/lept/' and then try again: the tesseract-operation works where it previously didn't.

You probably run into this problem if you run tesseract + leptonica as root first, then as another user.

**Comment by jbreiden** [comment]
This is a temporary, Linux specific Leptonica patch as a placeholder until upstream
does something official.

```c
--- leptonica/src/utils.c	2016-06-13 16:53:20.000000000 -0700
+++ leptonica/src/utils.c	2016-07-19 12:19:39.000000000 -0700
@@ -3152,6 +3152,10 @@
  *          (b) multiple threads from a single process call this function, or
  *          (c) there is the possibility of an attack where the intruder
  *              is logged onto the server and might try to guess filenames.
+ *
+ * IMPORTANT: This function has been modified as a temporary Debian
+ * specific workaround for Debian Bug #830660, while upstream works on
+ * an official solution. All input parameters are disabled.
  * </pre>
  */
 char *
@@ -3160,42 +3164,23 @@
                 l_int32      usetime,
                 l_int32      usepid)
 {
-char     buf[256];
-char    *newpath;
-l_int32  i, buflen, usec, pid, emptytail;
-
+char    *pathout;
+l_int32  size;
     PROCNAME("genTempFilename");
-
-    if (!dir)
-        return (char *)ERROR_PTR("dir not defined", procName, NULL);
-    if (dir && strlen(dir) == 1 && dir[0] == '/')
-        return (char *)ERROR_PTR("dir == '/' not permitted", procName, NULL);
-    if (tail && strlen(tail) > 0 && stringFindSubstr(tail, "/", NULL))
-        return (char *)ERROR_PTR("tail can't contain '/'", procName, NULL);
-    emptytail = tail && (strlen(tail) == 0);
-    if (!usetime && !usepid && (!tail || emptytail))
-        return (char *)ERROR_PTR("name can't be a directory", procName, NULL);
-
-    if (usepid) pid = getpid();
-    buflen = sizeof(buf);
-    for (i = 0; i < buflen; i++)
-        buf[i] = 0;
-    l_getCurrentTime(NULL, &usec);
-
-    newpath = genPathname(dir, NULL);
-    if (usetime && usepid)
-        snprintf(buf, buflen, "%s/%d_%d_", newpath, usec, pid);
-    else if (usetime)
-        snprintf(buf, buflen, "%s/%d_", newpath, usec);
-    else if (usepid)
-        snprintf(buf, buflen, "%s/%d_", newpath, pid);
-    else
-        snprintf(buf, buflen, "%s/", newpath);
-    LEPT_FREE(newpath);
-
-    return stringJoin(buf, tail);
+#ifdef _WIN32
+  #error "mkstemp is probably not supported on Windows"
+#endif  /*  _WIN32 */
+    char pattern[] = "/tmp/lept.XXXXXX";
+    int fd = mkstemp(pattern);
+    if (fd == -1) {
+      return (char *)ERROR_PTR("failed to get tempfile", procName, NULL);
+    }
+    close(fd);
+    if ((pathout = (char *)LEPT_CALLOC(sizeof(pattern), sizeof(char))) == NULL)
+      return (char *)ERROR_PTR("pathout not made", procName, NULL);
+    strncpy(pathout, pattern, sizeof(pattern) - 1);
+    return pathout;
 }
-
 
 /*!
  * \brief   extractNumberFromFilename()
```

**Comment by ebogaard** [comment]
Thanks for the temporary patch.
I think this bug can be closed as it's really a leptonica problem, which is handled there.

**Comment by amitdo** [comment]
>This function has been modified as a temporary Debian
+ * specific workaround for Debian Bug #830660, while upstream works on
+ * an official solution.

@DanBloomberg
Do you remember this issue?

**Comment by DanBloomberg** [comment]
yes, I made some changes last week on the leptonica github head:

    https://github.com/danbloomberg/leptonica

which I believe, together with the use of the new function l_makeTempFilename(), will resolve the issue for tesseract.  This function was also implemented for windows; it compiles but has not been tested further.

I haven't yet made a new release of leptonica.  That awaits some testing on windows.
